ocaml/typing/jkind_types.mli

```diff
@@ -41,108 +41,124 @@
+      | Equal_mutated_first
+      | Equal_mutated_second
+      | Equal_no_mutation
 
-  val equate_tracking_mutation : t -> t -> equate_result
+    val equate_tracking_mutation : t -> t -> equate_result
 
-  val get : t -> t
+    val get : t -> t
 
-  val to_string : t -> string
-end
-
-module Layout : sig
-  type 'sort layout =
-    | Sort of 'sort
-    | Any
+    val to_string : t -> string
+  end
 
-  module Const : sig
-    type t = Sort.const layout
-
-    module Legacy : sig
-      type t =
-        | Any
-        | Value
-        | Void
-        | Immediate64
-        | Immediate
-        | Float64
-        | Float32
-        | Word
-        | Bits32
-        | Bits64
+  module Layout : sig
+    type 'sort layout =
+      | Sort of 'sort
+      | Any
+
+    module Const : sig
+      type t = Sort.const layout
+
+      module Legacy : sig
+        type t =
+          | Any
+          | Value
+          | Void
+          | Immediate64
+          | Immediate
+          | Float64
+          | Float32
+          | Word
+          | Bits32
+          | Bits64
+      end
     end
+
+    type t = Sort.t layout
   end
 
-  type t = Sort.t layout
-end
+  module Externality : sig
+    type t =
+      | External
+      | External64
+      | Internal
+  end
 
-module Externality : sig
-  type t =
-    | External
-    | External64
-    | Internal
-end
+  module Modes = Mode.Alloc.Const
+
+  module Jkind_desc : sig
+    type 'type_expr t =
+      { layout : Layout.t;
+        modes_upper_bounds : Modes.t;
+        externality_upper_bound : Externality.t
+      }
+  end
 
-module Modes = Mode.Alloc.Const
+  type 'type_expr history =
+    | Interact of
+        { reason : Jkind_intf.History.interact_reason;
+          lhs_jkind : 'type_expr Jkind_desc.t;
+          lhs_history : 'type_expr history;
+          rhs_jkind : 'type_expr Jkind_desc.t;
+          rhs_history : 'type_expr history
+        }
+    | Creation of Jkind_intf.History.creation_reason
 
-module Jkind_desc : sig
   type 'type_expr t =
-    { layout : Layout.t;
-      modes_upper_bounds : Modes.t;
-      externality_upper_bound : Externality.t
+    { jkind : 'type_expr Jkind_desc.t;
+      history : 'type_expr history;
+      has_warned : bool
     }
-end
 
-type 'type_expr history =
-  | Interact of
-      { reason : Jkind_intf.History.interact_reason;
-        lhs_jkind : 'type_expr Jkind_desc.t;
-        lhs_history : 'type_expr history;
-        rhs_jkind : 'type_expr Jkind_desc.t;
-        rhs_history : 'type_expr history
+  module Const : sig
+    type 'type_expr t =
+      { layout : Layout.Const.t;
+        modes_upper_bounds : Modes.t;
+        externality_upper_bound : Externality.t
       }
-  | Creation of Jkind_intf.History.creation_reason
+  end
+
+  type 'type_expr annotation = 'type_expr Const.t * Jane_syntax.Jkind.annotation
+end
 
-type 'type_expr t =
-  { jkind : 'type_expr Jkind_desc.t;
-    history : 'type_expr history;
-    has_warned : bool
+type 'a arrow =
```

Comment:
This surprises me. I think it's probably better wrapped in a module like other types. And why make it polymorphic this way?

Reply:
Sure, I can wrap it in a module.
It's polymorphic so it can be reused for both `Jkind` and `Jkind.Const` - it got placed here accidentally, I will revert it here and use the module wrap in a further PR.